Google.Api.Generator/Generation/ResourceNamesGenerator.cs

```diff
@@ -70,140 +70,402 @@ public ParamProperty(SourceFileContext ctx, string name)
+                        Return(false)
+                    )
+                    .WithXmlDoc(
+                        XmlDoc.Summary("Tries to parse a resource name in string form into a new ", _ctx.Type(_def.ContainerTyp), " instance."),
+                        XmlDoc.Remarks("To parse successfully the resource name must be one of the following:", xmlDocUl, "Or an ", _ctx.Type<UnknownResourceName>(), " if ", allowUnknown, " is ", true),
+                        XmlDoc.Param(name, "The resource name in string form. Must not be ", null, "."),
+                        XmlDoc.Param(allowUnknown, "If ", true, ", will successfully parse an unknown resource name into an ", _ctx.Type<UnknownResourceName>(), "."),
+                        XmlDoc.Param(result, "When this method returns, the parsed ", _ctx.Type(_def.ContainerTyp), ", or ", null, " if parsing fails."),
+                        XmlDoc.Returns(true, " if the name was parsed succssfully; othrewise ", false));
+            }
 
-                MethodDeclarationSyntax Parse(FieldDeclarationSyntax templateField)
+            private IEnumerable<MethodDeclarationSyntax> Froms()
+            {
+                foreach (var single in _def.Defs)
                 {
-                    var name = Parameter(_ctx.Type<string>(), def.FieldName);
-                    var resourceName = Local(_ctx.Type<TemplatedResourceName>(), "resourceName");
-                    return Method(Public | Static, _ctx.Type(def.One.ResourceNameTyp), "Parse")(name)
-                        .WithBody(
-                            _ctx.Type(typeof(GaxPreconditions)).Call(nameof(GaxPreconditions.CheckNotNull))(name, Nameof(name)),
-                            resourceName.WithInitializer(templateField.Call("ParseName")(name)),
-                            Return(New(_ctx.Type(def.One.ResourceNameTyp))(def.One.Template.ParameterNames.Select((_, i) => resourceName.ElementAccess(i)))))
+                    var name = Parameter(_ctx.Type(single.ResourceNameTyp), single.ResourceNameTyp.Name.ToLowerCamelCase());
+                    yield return Method(Public | Static, _ctx.Type(_def.ContainerTyp), "From")(name)
+                        .WithBody(New(_ctx.Type(_def.ContainerTyp))(_ctx.Type(OneOfTypeEnumTyp).Access(single.ResourceNameTyp.Name), name))
                         .WithXmlDoc(
-                            XmlDoc.Summary("Parses the given ", XmlDoc.C(def.DocName), " resource name in string form into a new ", _ctx.Type(def.One.ResourceNameTyp), " instance."),
-                            XmlDoc.Param(name, "The ", XmlDoc.C(def.DocName), " resource name in string form. Must not be ", null, "."),
-                            XmlDoc.Returns("The parsed ", _ctx.Type(def.One.ResourceNameTyp), " if successful.")
-                        );
+                            XmlDoc.Summary("Construct a new instance of ", _ctx.Type(_def.ContainerTyp), " from the provided ", _ctx.Type(single.ResourceNameTyp), "."),
```

Comment:
Construct => Constructs? Or is it "Construct" to match previous code?

Reply:
Done